### PR TITLE
Display machine learning demos from HF Spaces

### PR DESCRIPTION
**Pull request by aliabid94**
This pull request introduces a new [arXiv Labs](https://labs.arxiv.org/) integration that displays links to interactive machine learning demos from [Hugging Face Spaces](https://huggingface.co/spaces). It mirrors the implementation used for Replicate AI demos.

Screenshot:
Screen Shot 2022-05-08 at 10 09 08 PM

Demo:
arxiv-spaces-integrations

Running the development environment. From the root of this repository, run:
- `docker build . -t arxiv/browse`
- `docker run -it --publish 8000:8000 arxiv/browse`
- I had to change the Dockerfile line 23 to `RUN pip install -U pip pipenv==2021.5.29`. Not sure why, you may need to do this to run locally. 

Once the Docker is running:
- go to any paper, e.g. `http://localhost:8000/abs/0704.0988`
- Because none of the dummy dev papers are linked to any HF model, use the "override_paper_id" URL param to override the paper_id used to search for demos, e.g. `http://localhost:8000/abs/0704.0988?override_paper_id=1810.04805` or `http://localhost:8000/abs/0704.0988?override_paper_id=1912.08777 `
- ArXiv hides the demo tab unless the paper is a CS paper, which none of the dummy dev papers are, so run the follow in the Inspector: `document.querySelector("#labstabs-demos-label").style.display="block"; document.querySelector("#labstabs-demos-input").removeAttribute("disabled")
`
- Click the Demo tab and open the Huggingface Spaces toggle

### CODE REVIEW

File: browse/static/js/spaces.js [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+      }
+      const huggingfaceSpaceApi = `${huggingfaceApiHost}/spaces/${space.id}`
+      response = await fetch(huggingfaceSpaceApi);
+      if (!response.ok) {
+        console.error(`Unable to fetch data from ${huggingfaceSpaceApi}`);
+        return;
+      }
+      let space_data = await response.json();
+      spaces_data.push(space_data);
+    }));
```
[the remark this answers is not in the document]
Comment (aliabid94) [reply browse/static/js/spaces.js:64]: I wanted to get the model card color and emoji information, but I agree making this many API calls to the HF API is not ideal. huggingfaceSpacesFromModelsApi only returns a list of spaces as  
```
[{
  "id": str,
  "sdk": str,
  "likes": int,
  "models": [str]
}]
```
and nothing about card emoji / color.

[the remark this answers is not in the document]
Comment (julien-c) [reply browse/static/js/spaces.js:64]: you can pass `?full=true` to that API call though – i tagged you on the corresponding internal PR. Thanks!

[the remark this answers is not in the document]
Comment (aliabid94) [reply browse/static/js/spaces.js:64]: ah just saw it, thanks!

File: browse/static/js/spaces.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+          <span class="spaces-emoji">${model.cardData.emoji}</span>
+          <span class="spaces-title">${model.cardData.title}</span>
+        </a>
+        <div class="spaces-model-details">
+          <a href="${huggingfaceSpacesHost}/${model.id}">
+            <h3 class="spaces-model-details-heading">${model.id}</h3>
+          </a>
+          <p class="spaces-model-sdk">A ${model.cardData.sdk} demo.</p>
+        </div>
+      </div>
```
[the remark this answers is not in the document]
Comment (aliabid94) [reply browse/static/js/spaces.js:146]: I didn't realize there was a thumbnail for each space I could reference, I didn't see it in the hub API docs. Where can I access this?

[the remark this answers is not in the document]
Comment (julien-c) [reply browse/static/js/spaces.js:146]: you can just use this URL pattern: https://thumbnails.huggingface.co/social-thumbnails/spaces/julien-c/coqui.png

[the remark this answers is not in the document]
Comment (julien-c) [reply browse/static/js/spaces.js:146]: it's not documented but the URL is stable

File: browse/static/js/spaces.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+      console.error(`Unable to fetch model data from ${huggingfaceModelsFromPaperApi}`)
+      render([]);
+      return;
+    }
+    let models = await response.json();
+    if (models.length === 0) {
+      render([]);
+      return;
+    }
+    const model_ids = models.map(m => m.id).join(",");
```
[the remark this answers is not in the document]
Comment (abidlabs) [reply browse/static/js/spaces.js:42]: I played around with some sample API calls, and for some papers (e.g. DistilBERT), the number of Spaces can get really large. What about if you _only_ show the full thumbnail for the top demo (based on the number of likes) and then just list the other Spaces in a more condensed way?

[the remark this answers is not in the document]
Comment (abidlabs) [reply browse/static/js/spaces.js:42]: From a user perspective, if they see 20 different demos, they probably won't click on any of them

**Comment by abidlabs** [comment]
If there are no replicate demos, but there _are_ Spaces demos, it leads this contradictory message:

image

I don't know if we are allowed to touch the replicate code, but the phrase "no demos found..." should only appear if there are NO replicate demos AND no Spaces demos

Some nits:

- Instead of "A streamlit demo" or "A Gradio demo", how about "<Streamlit logo> streamlit" or "<Gradio logo> gradio" (I know there are other SDKs we might want to support, but at least for these more common ones, I think the logo would look much nicer than just the text)

- The information next to the demo is very bare. What about including the last modified date? It's available from the API endpoint and would be of interest to people I think

File: browse/static/js/spaces.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (julien-c) [reply browse/static/js/spaces.js:42]: BTW you should also be able to order by likes in the API calls by adding `sort=likes&direction=-1` to your url query params

[the remark this answers is not in the document]
Comment (julien-c) [reply browse/static/js/spaces.js:42]: (`limit` is also supported)

**Comment by aliabid94** [comment]
Can't access any images in the repo from JS because templating only applies only to HTML files, and don't want to hardcode any logo URLs. I tweaked the look a look a little, take a look. Also added "No Spaces / No Replicate demos" message.

> If there are no replicate demos, but there _are_ Spaces demos, it leads this contradictory message:
> 
> image
> 
> I don't know if we are allowed to touch the replicate code, but the phrase "no demos found..." should only appear if there are NO replicate demos AND no Spaces demos
> 
> Some nits:
> 
> * Instead of "A streamlit demo" or "A Gradio demo", how about " streamlit" or " gradio" (I know there are other SDKs we might want to support, but at least for these more common ones, I think the logo would look much nicer than just the text)
> * The information next to the demo is very bare. What about including the last modified date? It's available from the API endpoint and would be of interest to people I think

**Comment by aliabid94** [comment]
Implemented a "load more" button (using CSS only because event listeners / JS are sanitized out) as well. Displays 10 models at first, and the rest on click.

**Comment by abidlabs** [comment]
Tested and looks great to me from the Hugging Face end @aliabid94

**Pull request closed without merging**